### PR TITLE
fix(upstream): Fix discrepancy with upstream

### LINKED ISSUE #49: Ensure equivalence with upstream
The `addresses` mod differs from the official TS SDK after recent upstream updates. For example the [`swap_router02_address`](https://github.com/malik672/uniswap-sdk-core-rust/blob/master/src/addresses.rs#L487) function should be changed after https://github.com/Uniswap/sdk-core/pull/120. Same for `V2_FACTORY_ADDRESSES` after https://github.com/Uniswap/sdk-core/pull/115. Also `V2_ROUTER_ADDRESSES` was modified in https://github.com/Uniswap/sdk-core/pull/113.

### PR DESCRIPTION
**Pull request by malik672**
added upstream equivalence

### CODE REVIEW

File: src/addresses.rs [hunk 1]
```diff
@@ -57,25 +57,91 @@ pub const UNISWAP_NFT_AIRDROP_CLAIM_ADDRESS: Address =
...
+            ChainId::MAINNET as u64,
+            address!("5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"),
+        );
+        m.insert(
+            ChainId::GOERLI as u64,
+            address!("5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"),
+        );
+        m.insert(
+            ChainId::SEPOLIA as u64,
+            address!("5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
            address!("B7f907f7A9eBC822a80BD25E224be42Ce0A698A0"),
```

File: src/addresses.rs [hunk 2]
```diff
@@ -57,25 +57,91 @@ pub const UNISWAP_NFT_AIRDROP_CLAIM_ADDRESS: Address =
...
-            ChainId::BNB,
-            ChainId::AVALANCHE,
-            ChainId::BASE,
-        ]
-    );
+    pub static ref V2_FACTORY_ADDRESSES: HashMap<u64, Address> = {
+        let mut m = HashMap::new();
+        m.insert(
+            ChainId::MAINNET as u64,
+            address!("5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
            V2_FACTORY_ADDRESS,
```

File: src/addresses.rs [hunk 3]
```diff
@@ -57,25 +57,91 @@ pub const UNISWAP_NFT_AIRDROP_CLAIM_ADDRESS: Address =
...
+        );
+        m.insert(
+            ChainId::GOERLI as u64,
+            address!("5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f"),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
            V2_FACTORY_ADDRESS,
```

File: src/addresses.rs [hunk 4]
```diff
@@ -57,25 +57,91 @@ pub const UNISWAP_NFT_AIRDROP_CLAIM_ADDRESS: Address =
...
 pub const V2_ROUTER_ADDRESS: Address = address!("7a250d5630B4cF539739dF2C5dAcb4c659F2488D");
 
 lazy_static! {
-    pub static ref V2_ROUTER_ADDRESSES: AddressMap =
-        construct_same_address_map(V2_ROUTER_ADDRESS, &[]);
+    pub static ref V2_ROUTER_ADDRESSES: HashMap<u64, Address> = {
+        let mut m = HashMap::new();
+        m.insert(
+            ChainId::MAINNET as u64,
+            address!("7a250d5630B4cF539739dF2C5dAcb4c659F2488D"),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
            V2_ROUTER_ADDRESS,
```

File: src/addresses.rs [hunk 5]
```diff
@@ -57,25 +57,91 @@ pub const UNISWAP_NFT_AIRDROP_CLAIM_ADDRESS: Address =
...
+        );
+        m.insert(
+            ChainId::GOERLI as u64,
+            address!("7a250d5630B4cF539739dF2C5dAcb4c659F2488D"),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
            V2_ROUTER_ADDRESS,
```

File: src/addresses.rs [hunk 6]
```diff
@@ -97,38 +163,47 @@ impl Default for ChainAddresses {
     }
 }
 
-pub fn mainnet_address() -> ChainAddresses {
-    ChainAddresses {
-        v1_mixed_route_quoter_address: Some(address!("84E44095eeBfEC7793Cd7d5b57B7e401D7f1cA2E")),
-        ..Default::default()
-    }
+lazy_static! {
+    pub static ref MAINNET_ADDRESS: ChainAddresses = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref MAINNET_ADDRESSES: ChainAddresses = {
```

File: src/addresses.rs [hunk 7]
```diff
@@ -97,38 +163,47 @@ impl Default for ChainAddresses {
...
+    };
 }
 
-pub fn goerli_address() -> ChainAddresses {
-    ChainAddresses {
-        v1_mixed_route_quoter_address: Some(address!("Ba60b6e6fF25488308789E6e0A65D838be34194e")),
-        ..Default::default()
-    }
+lazy_static! {
+    pub static ref GOERLI_ADDRESS: ChainAddresses = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref GOERLI_ADDRESSES: ChainAddresses = {
```

File: src/addresses.rs [hunk 8]
```diff
@@ -97,38 +163,47 @@ impl Default for ChainAddresses {
...
-
-pub fn polygon_addresses() -> ChainAddresses {
-    ChainAddresses::default()
+lazy_static! {
+    pub static ref POLYGON_ADDRESSES: ChainAddresses = ChainAddresses::default();
 }
 
 /// celo v3 addresses
-pub fn celo_addresses() -> ChainAddresses {
+pub const CELO_ADDRESSES: ChainAddresses = {
```
Comment (shuhuiluo) [inline src/addresses.rs]: Redundant `{`.

File: src/addresses.rs [hunk 9]
```diff
@@ -205,10 +280,26 @@ pub fn arbitrum_goerli_addresses() -> ChainAddresses {
...
+pub const ARBITRUM_SEPOLIA_ADDRESSES: ChainAddresses = {
+    ChainAddresses {
+        v3_core_factory_address: address!("248AB79Bbb9bC29bB72f7Cd42F17e054Fc40188e"),
+        multicall_address: address!("2B718b475e385eD29F56775a66aAB1F5cC6B2A0A"),
+        quoter_address: address!("2779a0CC1c3e0E44D2542EC3e79e3864Ae93Ef0B"),
+        v3_migrator_address: Some(address!("398f43ef2c67B941147157DA1c5a868E906E043D")),
+        nonfungible_position_manager_address: Some(address!(
+            "6b2937Bde17889EDCf8fbD8dE31C3C2a70Bc4d65"
+        )),
+        tick_lens_address: Some(address!("b52429333da969a0C79a60930a4Bf0020E5D1DE8")),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
        tick_lens_address: Some(address!("0fd18587734e5C2dcE2dccDcC7DD1EC89ba557d9")),
```

File: src/addresses.rs [hunk 10]
```diff
@@ -269,32 +360,81 @@ pub fn base_goerli_addresses() -> ChainAddresses {
...
+    ChainAddresses {
+        v3_core_factory_address: address!("aF37EC98A00FD63689CF3060BF3B6784E00caD82"),
+        multicall_address: address!("996a9858cDfa45Ad68E47c9A30a7201E29c6a386"),
+        quoter_address: address!("b51727c996C68E60F598A923a5006853cd2fEB31"),
+        v3_migrator_address: Some(address!("16678977CA4ec3DAD5efc7b15780295FE5f56162")),
+        nonfungible_position_manager_address: Some(address!(
+            "9d9386c042F194B460Ec424a1e57ACDE25f5C4b1"
+        )),
+        tick_lens_address: Some(address!("55B9dF5bF68ADe972191a91980459f48ecA16afC")),
+        swap_router02_address: None,
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
        swap_router02_address: Some(address!("0B14ff67f0014046b4b99057Aec4509640b3947A")),
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
Missing `V3_CORE_FACTORY_ADDRESSES`,`V3_MIGRATOR_ADDRESSES`, etc.

**Comment by malik672** [comment]
> Missing `V3_CORE_FACTORY_ADDRESSES`,`V3_MIGRATOR_ADDRESSES`, etc.

can you drop the link to them

**Comment by shuhuiluo** [comment]
> > Missing `V3_CORE_FACTORY_ADDRESSES`,`V3_MIGRATOR_ADDRESSES`, etc.
> 
> can you drop the link to them

https://github.com/Uniswap/sdk-core/blob/main/src/addresses.ts#L250

**Comment by malik672** [comment]
didn't think i missed some parts

File: src/addresses.rs (hunk 8 above)
Comment (malik672) [reply src/addresses.rs]: true, but my rust-analyzer has this red stuff near the closing bracket

File: src/addresses.rs [hunk 11]
```diff
@@ -486,7 +626,13 @@ pub fn v1_mixed_route_quoter_address() -> ChainAddress {
 }
 
 pub fn swap_router02_address(chain_id: u64) -> Address {
```
Comment (shuhuiluo) [inline src/addresses.rs:618]: Convert all addresses to lazy statics and upper case.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/addresses.rs [hunk 12]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-        v3_migrator_address: Some(address!("729004182cF005CEC8Bd85df140094b6aCbe8b15")),
-        nonfungible_position_manager_address: Some(address!(
-            "1238536071E1c677A632429e3655c799b22cDA52"
-        )),
-        tick_lens_address: Some(address!("d7f33bcdb21b359c8ee6f0251d30e94832baad07")),
-        swap_router02_address: None,
-        v1_mixed_route_quoter_address: None,
-    }
-}
+pub const SEPOLIA_ADDRESS: ChainAddresses = ChainAddresses {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
pub const SEPOLIA_ADDRESSES: ChainAddresses = ChainAddresses {
```

File: src/addresses.rs [hunk 13]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
+    v3_core_factory_address: address!("0227628f3F023bb0B980b67D528571c95c6DaC1c"),
+    multicall_address: address!("D7F33bCdb21b359c8ee6F0251d30E94832baAd07"),
+    quoter_address: address!("Ed1f6473345F45b75F8179591dd5bA1888cf2FB3"),
+    v3_migrator_address: Some(address!("729004182cF005CEC8Bd85df140094b6aCbe8b15")),
+    nonfungible_position_manager_address: Some(address!(
+        "1238536071E1c677A632429e3655c799b22cDA52"
+    )),
+    tick_lens_address: Some(address!("d7f33bcdb21b359c8ee6f0251d30e94832baad07")),
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    tick_lens_address: Some(address!("0b343475d44EC2b4b8243EBF81dc888BF0A14b36")),
```
https://github.com/Uniswap/sdk-core/issues/127

File: src/addresses.rs [hunk 14]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-                .unwrap()
-                .v3_core_factory_address,
-        );
-    }
-    chain_add
+lazy_static! {
+    /// V3 Contract Addresses
+    pub static ref V3_CORE_FACTORY_ADDRESSES: ChainAddress = {
+        let mut chain_add = ChainAddress::new();
+        for memo in SUPPORTED_CHAINS {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
        for chain_id in SUPPORTED_CHAINS {
```
Rename all `memo` which was Copilot's blunder to `chain_id`.

File: src/addresses.rs [hunk 15]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-    let mut chain_add = ChainAddress::new();
-    for memo in SUPPORTED_CHAINS {
-        if CHAIN_TO_ADDRESSES_MAP
-            .get(&(memo as u64))
-            .unwrap()
-            .nonfungible_position_manager_address
-            .is_some()
-        {
+lazy_static! {
+    pub static ref QUOTER_ADDRESS: ChainAddress = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref QUOTER_ADDRESSES: ChainAddress = {
```

File: src/addresses.rs [hunk 16]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-    }
-    chain_add
+        chain_add
+    };
 }
 
-pub fn ens_resgister_address_map() -> AddressMap {
-    construct_same_address_map(address!("00000000000C2E074eC69A0dFb2997BA6C7d2e1e"), &[])
+lazy_static! {
+    pub static ref NONFUNGIBLE_POSITION_MANAGER_ADDRESS: ChainAddress = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref NONFUNGIBLE_POSITION_MANAGER_ADDRESSES: ChainAddress = {
```

File: src/addresses.rs [hunk 17]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
 
-pub fn socks_controller_addresses() -> AddressMap {
-    let mut new_map = AddressMap::new();
-    new_map.insert(
-        ChainId::MAINNET as u64,
-        address!("65770b5283117639760beA3F867b69b3697a91dd"),
-    );
-    new_map
+lazy_static! {
+    pub static ref ENS_RESGISTER_ADDRESS_MAP: AddressMap =
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref ENS_REGISTRAR_ADDRESSES: AddressMap =
```

File: src/addresses.rs [hunk 18]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-            chain_add.insert(
-                memo as u64,
-                CHAIN_TO_ADDRESSES_MAP
-                    .get(&(memo as u64))
-                    .unwrap()
-                    .v1_mixed_route_quoter_address
-                    .unwrap(),
-            );
+lazy_static! {
+    pub static ref V1_MIXED_ROUTE_QUOTER_ADDRESS: ChainAddress = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref MIXED_ROUTE_QUOTER_V1_ADDRESSES: ChainAddress = {
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Comment by malik672** [comment]
@shuhuiluo check the failed test, can't discover the issue

File: src/addresses.rs [hunk 19]
```diff
@@ -552,7 +552,7 @@ lazy_static! {
 }
 
 lazy_static! {
-    pub static ref ENS_RESGISTER_ADDRESS_MAP: AddressMap =
+    pub static ref ENS_RESGISTER_ADDRESS: AddressMap =
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref ENS_REGISTRAR_ADDRESSES: AddressMap =
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
Fix `bigdecimal` to `0.4.2` to fix the tests.

File: src/addresses.rs [hunk 20]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-    let mut chain_add = ChainAddress::new();
-    for memo in SUPPORTED_CHAINS {
-        if CHAIN_TO_ADDRESSES_MAP
-            .get(&(memo as u64))
-            .unwrap()
-            .nonfungible_position_manager_address
-            .is_some()
-        {
+lazy_static! {
+    pub static ref QUOTER_ADDRESSESES: ChainAddress = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref QUOTER_ADDRESSES: ChainAddress = {
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/addresses.rs [hunk 21]
```diff
@@ -97,396 +151,478 @@ impl Default for ChainAddresses {
...
-            chain_add.insert(
-                memo as u64,
-                CHAIN_TO_ADDRESSES_MAP
-                    .get(&(memo as u64))
-                    .unwrap()
-                    .v1_mixed_route_quoter_address
-                    .unwrap(),
-            );
+lazy_static! {
+    pub static ref V1_MIXED_ROUTE_QUOTER_ADDRESSES: ChainAddress = {
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
    pub static ref MIXED_ROUTE_QUOTER_V1_ADDRESSES: ChainAddress = {
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Comment by malik672** [comment]
> a97a306
that's what it's currently at

**Comment by shuhuiluo** [comment]
> @shuhuiluo check the failed test, can't discover the issue

```
bigdecimal = "=0.4.2"
```

**Review by shuhuiluo — APPROVED** [review]
Squash and merge.

**Pull request merged**